cluster/gce/upgrade.sh

```diff
@@ -439,6 +439,75 @@ function do-node-upgrade() {
   echo "== Finished upgrading nodes to ${KUBE_VERSION}. ==" >&2
 }
 
+CURRENT_COREDNS_VERSION=$(${KUBE_ROOT}/cluster/kubectl.sh -n kube-system get deployment coredns -o=jsonpath='{$.spec.template.spec.containers[:1].image}' | cut -d ":" -f 2)
+
+function update-coredns-config() {
+  # Get the current CoreDNS version
+  local -r coredns_addon_path="/etc/kubernetes/addons/0-dns/coredns"
+  local -r tmpdir=/tmp
+  local -r download_dir=$(mktemp --tmpdir=${tmpdir} -d coredns-migration.XXXXXXXXXX)
+
+  # Get the new installed CoreDNS version
+  sleep 10
+  NEW_COREDNS_VERSION=$(${KUBE_ROOT}/cluster/kubectl.sh -n kube-system get deployment coredns -o=jsonpath='{$.spec.template.spec.containers[:1].image}' | cut -d ":" -f 2)
+  echo ${NEW_COREDNS_VERSION}
+
+  case "$(uname -m)" in
+      x86_64*)
+        host_arch=amd64
+        ;;
+      i?86_64*)
+        host_arch=amd64
+        ;;
+      amd64*)
+        host_arch=amd64
+        ;;
+      aarch64*)
+        host_arch=arm64
+        ;;
+      arm64*)
+        host_arch=arm64
+        ;;
+      arm*)
+        host_arch=arm
+        ;;
+      i?86*)
+        host_arch=x86
+        ;;
+      s390x*)
+        host_arch=s390x
+        ;;
+      ppc64le*)
+        host_arch=ppc64le
+        ;;
+      *)
+        echo "Unsupported host arch. Must be x86_64, 386, arm, arm64, s390x or ppc64le." >&2
+        exit 1
+        ;;
+    esac
+
+  # Download the CoreDNS migration tool
+  echo "Downloading the CoreDNS migration tool"
+  wget -P ${download_dir} "https://github.com/coredns/corefile-migration/releases/download/migration/v1.0.1/corefile-tool-${host_arch}" >/dev/null 2>&1
+  chmod +x ${download_dir}/corefile-tool-${host_arch}
+
+  # Migrate the CoreDNS ConfigMap depending on whether it is being downgraded or upgraded.
+  ${KUBE_ROOT}/cluster/kubectl.sh -n kube-system get cm coredns -o jsonpath='{.data.Corefile}' > ${download_dir}/Corefile-old
+
+  if test "$(printf '%s\n' ${CURRENT_COREDNS_VERSION} ${NEW_COREDNS_VERSION} | sort -V | head -n 1)" != ${NEW_COREDNS_VERSION}; then
+     echo "Upgrading the CoreDNS ConfigMap"
+     ./corefile-tool-${host_arch} migrate --from ${CURRENT_COREDNS_VERSION} --to ${NEW_COREDNS_VERSION} --corefile ${download_dir}/Corefile-old > ${download_dir}/Corefile
+     ${KUBE_ROOT}/cluster/kubectl.sh -n kube-system create configmap coredns --from-file ${download_dir}/Corefile -o yaml --dry-run | ${KUBE_ROOT}/cluster/kubectl.sh replace -f -
+  else
+     echo "Applying the latest default CoreDNS configuration."
+     gcloud compute --project ${PROJECT}  ssh --zone ${ZONE} ${MASTER_NAME} --command "cat ${coredns_addon_path}/coredns.yaml" > ${download_dir}/coredns-manifest.yaml
```

Comment:
`gcloud scp` may be better for coping file?

Reply:
yes! done.